android/modules/ui/src/java/ti/modules/titanium/ui/ScrollableViewProxy.java

```diff
@@ -190,40 +121,107 @@ public boolean handleMessage(Message msg)
 		return handled;
 	}
 
-	@Kroll.getProperty
-	public Object getViews()
+	public void removeAllViews()
 	{
-		TiViewProxy[] childViewArray = new TiViewProxy[0];
-		TiUIScrollableView view = getView();
-		if (view != null) {
-			childViewArray = view.getViews().toArray(childViewArray);
+		for (final TiViewProxy view : this.views) {
+			view.releaseViews();
+			view.setParent(null);
+		}
+		this.views.clear();
+
+		if (scrollableView != null) {
+			scrollableView.getAdapter().notifyDataSetChanged();
 		}
-		return childViewArray;
+	}
+
+	public ArrayList<TiViewProxy> getViewsList()
+	{
+		return (ArrayList<TiViewProxy>) this.views;
+	}
+
+	@Kroll.getProperty
+	public TiViewProxy[] getViews()
+	{
+		return this.views.toArray(new TiViewProxy[0]);
 	}
 
 	@Kroll.setProperty
-	public void setViews(Object viewsObject)
+	public void setViews(Object views)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_SET_VIEWS), viewsObject);
+		removeAllViews();
+
+		if (views instanceof Object[]) {
+			for (final Object view : (Object[]) views) {
+				if (view instanceof TiViewProxy) {
+					addView((TiViewProxy) view);
+				}
+			}
+		}
 	}
 
 	@Kroll.method
-	public void addView(Object viewObject)
+	public void addView(TiViewProxy view)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_ADD_VIEW), viewObject);
+		if (!this.views.contains(view)) {
+			view.setActivity(getActivity());
+			view.setParent(this);
+			this.views.add(view);
+
+			if (scrollableView != null) {
+				scrollableView.getAdapter().notifyDataSetChanged();
+			}
+		}
 	}
 
 	@Kroll.method
 	public void insertViewsAt(int insertIndex, Object viewObject)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_INSERT_VIEWS_AT, insertIndex, 0),
-											viewObject);
+		if (viewObject instanceof TiViewProxy) {
+			final TiViewProxy view = (TiViewProxy) viewObject;
+
+			if (!this.views.contains(view)) {
+				view.setActivity(getActivity());
+				view.setParent(this);
+				this.views.add(insertIndex, view);
+			}
+
+		} else if (viewObject instanceof Object[]) {
+			Object[] views = (Object[]) viewObject;
+
+			for (int i = 0; i < views.length; i++) {
+				insertViewsAt(insertIndex, views[i]);
+			}
+		} else {
+
+			// Skip updating adapter.
+			return;
+		}
+
+		if (scrollableView != null) {
+			scrollableView.getAdapter().notifyDataSetChanged();
+		}
 	}
 
 	@Kroll.method
 	public void removeView(Object viewObject)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_REMOVE_VIEW), viewObject);
+		if (viewObject instanceof Number) {
+			this.views.remove(viewObject);
```

Comment:
I'm pretty sure you need to cast `viewObject` as `(int)` so that it will call the [remove(int)](https://developer.android.com/reference/java/util/ArrayList#remove(int)) method instead of the [remove(Object)](https://developer.android.com/reference/java/util/ArrayList#remove(java.lang.Object)) method.

Reply:
The `ScrollableViewAddRemoveTest.js` script I attached to [TIMOB-28459](https://jira.appcelerator.org/browse/TIMOB-28459) proves that calling `ScrollableView.removeView()` by index is now broken. When running this script, tap on the "Remove" button and select "Current" page from the dialog to remove the currently displayed page by index. The script's remove "First" and "Last" buttons remove by reference and work correctly.